src/Microsoft.Identity.Web/Resource/TokenAuthorizationRules.cs

```diff
@@ -0,0 +1,150 @@
+            }
+
+            // Does the web API demand specific clients (oid)
+            if (tokenAuthorizationOptions.AllowedClientApplications != null)
+            {
+                VerifyClaimIsInAcceptedValues(
+                    claimsPrincipal,
+                    new[] { ClaimConstants.Oid, ClaimConstants.ObjectId },
+                    tokenAuthorizationOptions.AllowedClientApplications,
+                    IDWebErrorMessage.ApiAllowedClientApplicationsCantBeComputed,
+                    IDWebErrorMessage.ApiClientApplicationNotAllowed);
+            }
+        }
+
+        /// <summary>
+        /// Verify the value of a claim.
+        /// </summary>
+        /// <param name="claimsPrincipal">Claims.</param>
+        /// <param name="claimName">Name (type) of the claim to verify.</param>
+        /// <param name="expectedValue">Expected value.</param>
+        /// <param name="messageWhenClaimsNotFound">Error message thrown when the claim is not found.</param>
+        /// <param name="messageWhenUnexpectedClaimValue">Error message thrown when claim value is not the expected one.</param>
+        private static void VerifyClaimEquals(
+            ClaimsPrincipal claimsPrincipal,
+            string claimName,
+            string expectedValue,
+            string messageWhenClaimsNotFound,
+            string messageWhenUnexpectedClaimValue)
+        {
+            string? claimValue = ClaimsPrincipalExtensions.GetClaimValue(claimsPrincipal, claimName);
+            if (claimValue == null)
+            {
+                throw new UnauthorizedAccessException(messageWhenClaimsNotFound);
+            }
+
+            if (string.Equals(claimValue, expectedValue, StringComparison.OrdinalIgnoreCase))
+            {
+                throw new UnauthorizedAccessException(messageWhenUnexpectedClaimValue);
+            }
+        }
+
+        /// <summary>
+        /// Verify the value of a claim.
+        /// </summary>
+        /// <param name="claimsPrincipal">Claims.</param>
+        /// <param name="claimName">Name (type) of the claim to verify.</param>
+        /// <param name="expectedValue">Expected value.</param>
+        /// <param name="messageWhenClaimsNotFound">Error message thrown when the claim is not found.</param>
+        /// <param name="messageWhenUnexpectedClaimValue">Error message thrown when claim value is not the expected one.</param>
+        private static void VerifyClaimDoesNotExistOrDoesNotEquals(
+            ClaimsPrincipal claimsPrincipal,
+            string claimName,
+            string expectedValue,
+            string messageWhenClaimsNotFound,
+            string messageWhenUnexpectedClaimValue)
+        {
+            string? claimValue = ClaimsPrincipalExtensions.GetClaimValue(claimsPrincipal, claimName);
+            if (claimValue != null)
+            {
+                if (!string.Equals(claimValue, expectedValue, StringComparison.OrdinalIgnoreCase))
```

Comment:
This should throw if the value _is_ equal to `expectedValue`, right? (Also, that parameter should probably be named something else, since here we're checking that the value is _not_ this?)